sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,221 @@
+    device_config |= (1 << HMAC_CFG_DIGEST_SWAP);
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == kDifHmacModeHmac) {
+    device_config |= (1 << HMAC_CFG_HMAC_EN);
+
+    // Set the key.
+    for (int i = 0; i < HMAC_KEY_REG_COUNT; i++) {
+      ptrdiff_t offset = i * sizeof(uint32_t);
+      mmio_region_write32(config->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                          config->key[i]);
+    }
+  }
+  device_config |= (1 << HMAC_CFG_SHA_EN);
+
+  // Write the configuration.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, device_config);
+
+  // Begin SHA or HMAC operation.
+  mmio_region_nonatomic_set_bit32(config->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+
+  // Setup state tracking.
+  hmac->base_addr = config->base_addr;
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_fifo_error_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
+                                         const void *data, size_t len,
+                                         size_t *bytes_sent) {
+  if (!hmac || !data) {
+    return kDifHmacFifoInvalidArgument;
+  }
+
+  const uint8_t *data_sent = (uint8_t *)data;
+  int words_remaining = len / sizeof(uint32_t);
+  int bytes_remaining = len % sizeof(uint32_t);
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word;
+
+    memcpy(&word, data_sent, sizeof(uint32_t));
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    words_remaining--;
+  }
+
+  // Once all words are written write remaining bytes individually.
+  while (words_remaining == 0 && bytes_remaining > 0 &&
+         get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    bytes_remaining--;
+  }
+
+  if (bytes_sent) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
+  }
+
+  if (words_remaining > 0 || bytes_remaining > 0) {
+    return kDifHmacFifoFull;
+  }
+  return kDifHmacFifoNoError;
+}
+
+dif_hmac_error_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
+                                             uint32_t *fifo_depth) {
+  if (!hmac || !fifo_depth) {
+    return kDifHmacInvalidArgument;
+  }
+
+  *fifo_depth = get_fifo_entry_count(hmac);
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_get_msg_len(const dif_hmac_t *hmac,
+                                      uint64_t *msg_len) {
+  if (!hmac || !msg_len) {
+    return kDifHmacInvalidArgument;
+  }
+
+  uint64_t msg_lower =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_LOWER_REG_OFFSET);
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER_REG_OFFSET);
+
+  *msg_len = (msg_upper << (sizeof(uint32_t) * 8)) | msg_lower;
```

Comment:
Would replacing `8` with `CHAR_BIT` be an acceptable replacement? I prefer keeping `sizeof()` instead of just a constant since it's more clear where the value comes from.

Reply:
I take my comment back since this is mostly a matter of style and we may eventually introduce a utility function for this operation.